game-core/src/main/java/games/strategy/triplea/ai/pro/util/ProMoveUtils.java

```diff
@@ -411,40 +401,34 @@ public static void calculateBombingRoutes(
                   u,
                   player);
         }
-        moveRoutes.add(route);
+        moves.add(new MoveDescription(unitList, route));
       }
     }
-  }
-
-  public static void doMove(
-      final List<Collection<Unit>> moveUnits,
-      final List<Route> moveRoutes,
-      final IMoveDelegate moveDel) {
-    doMove(moveUnits, moveRoutes, null, moveDel);
+    return moves;
   }
 
   /**
    * Moves the specified groups of units along the specified routes, possibly using the specified
    * transports.
    */
-  public static void doMove(
-      final List<Collection<Unit>> moveUnits,
-      final List<Route> moveRoutes,
-      final List<Map<Unit, Unit>> transportsToLoad,
-      final IMoveDelegate moveDel) {
-
+  public static void doMove(final List<MoveDescription> moves, final IMoveDelegate moveDel) {
     final GameData data = ProData.getData();
 
     // Group non-amphib units of the same type moving on the same route
-    if (transportsToLoad == null) {
-      for (int i = 0; i < moveRoutes.size(); i++) {
-        final Route r = moveRoutes.get(i);
-        for (int j = i + 1; j < moveRoutes.size(); j++) {
-          final Route r2 = moveRoutes.get(j);
+    // TODO: #5499 Use MoveBatcher here - or ideally at the time the moves are being generated.
+    final boolean noTransportLoads =
+        moves.stream().allMatch(move -> move.getUnitsToTransports().isEmpty());
+    if (noTransportLoads) {
+      for (int i = 0; i < moves.size(); i++) {
+        final Route r = moves.get(i).getRoute();
+        for (int j = i + 1; j < moves.size(); j++) {
+          final Route r2 = moves.get(j).getRoute();
           if (r.equals(r2)) {
-            moveUnits.get(j).addAll(moveUnits.get(i));
-            moveUnits.remove(i);
-            moveRoutes.remove(i);
+            final var mergedUnits = new ArrayList<Unit>();
+            mergedUnits.addAll(moves.get(j).getUnits());
+            mergedUnits.addAll(moves.get(i).getUnits());
+            moves.set(i, new MoveDescription(mergedUnits, r));
+            moves.remove(i);
```

Comment:
I don't think I understand the intent of these two lines. We set element 'i', then we remove element 'i'. What's the effect of the 'set'?

Reply:
@asvitkine please double check this one, I think it might supposed to be `remove(j)`

Running through a debugger, I noticed 'move' `i` gets the aggregation, then we it gets removed. It looks like it would make sense if it were `j` being removed.

Example below, with 'move @ index 2' being merged with 'move @ index 3'

![Screenshot from 2019-11-21 20-58-51](https://user-images.githubusercontent.com/12397753/69399417-e71e9d80-0ca2-11ea-9bdc-6185f4f8add8.png)
![Screenshot from 2019-11-21 21-00-00](https://user-images.githubusercontent.com/12397753/69399418-e7b73400-0ca2-11ea-996a-c351bd0c9b5a.png)
![Screenshot from 2019-11-21 21-01-47](https://user-images.githubusercontent.com/12397753/69399419-e7b73400-0ca2-11ea-970b-d673954f9203.png)

Reply:
Thanks for noticing, this was indeed a bug.

It was supposed to be set(j), not i. It's meant to match the previous logic, which was:

```
            moveUnits.get(j).addAll(moveUnits.get(i));
            moveUnits.remove(i);
            moveRoutes.remove(i);
```

i.e. making entry at index j be the merged move and removing i.

Fixed.